blog/_posts/2018-04-12-in-a-nutshell.md

```diff
@@ -0,0 +1,93 @@
+
+Then, we need to define and implement semantic operations such as
+
+ - what are the members that can be selected on this expression?
+ - which subclasses are defined for a sealed trait?
+ - does this expression conform to some expected type?
+
+Finally, we need to connect the new reflection layer to the existing high-level macro system. This looks not very difficult. In essence, we need to define a pair of mappings between high level trees of type `scala.quoted.Expr[T]` and lower-level Tasty trees of type `tasty.Term`. Mapping a high-level tree to a low-level one simply means exposing its structure. Mapping a a low-level tree to a high-level tree of type `scala.quoted.Expr[T]` means checking that the low-level tree has indeed the given type `T`. That should be all.
+
+## Future Macros
+
+Adopting this scheme gives already some idea what Scala 3 macros will look like. First, they will run after the typechecking phase is finished because that is when Tasty trees are generated and consumed. This means macros will be blackbox - a macro expansion cannot influence the type of the expanded expression as seen from the typechecker. A long as that constraint is satisfied we should be able to support both `def` macros and annotation macros.
+
+We might support some forms of whitebox macros by allowing macros in the types themselves. These macros would be highlevel only, and would integrate with implicit search. A sketch of such as system is outlined in [Dotty PR 3844](https://github.com/lampepfl/dotty/pull/3844]).
+
+The Scala 3 language should also directly incorporate some constructs that so far required
+advanced macro code to define. For instance:
+
+ - We model lazy implicits directly using
+[by-name parameters](http://dotty.epfl.ch/docs/reference/implicit-by-name-parameters.html) instead of through a complicated macro.
+ - Native [type lambdas](http://dotty.epfl.ch/docs/reference/type-lambdas.html) reduce the need for
+   [kind projector](https://github.com/non/kind-projector).
+ - We also plan to offer the basic building blocks for [generic programming](https://github.com/milessabin/shapeless) and [typeclass derivation](https://github.com/propensive/magnolia).
+
+## Please Give Us Your Feedback!
+
+What do you think of the macro roadmap? Your feedback would be much appreciated. There is also lots of scope to shape the future by contributing to the development.
```

Comment:
Where is best for contributors to join the discussion? Would be good to include link.

Reply:
I was wondering about this as well. Should we point them to scala contributors or open the discussion on Disquus? Do we still support blog discussions on the site?

Reply:
Definitely not disqus. Please create a thread on contributors. Personally, I think the whole article would make more sense as a contributors post. Posting on scala-lang, which we mostly use for release announcements, gives it the same air of doneness. We know the difference, but does a casual reader of the website?

Reply:
I think Scala contributors is the best place for these discussions. I think you can post an "Unlisted" thread to get a link without public post
<img width="538" alt="screen shot 2018-04-11 at 11 30 36" src="https://user-images.githubusercontent.com/1408093/38608489-cf0a646a-3d7b-11e8-8027-2c6c858fe07b.png">

Reply:
I am open to either contributors or the blog. But the style of the blog is more oriented towards users, not contributors. What do others prefer?

Reply:
I would expect most macro authors to be on contributors. They really should be, at least, since they're basically contributing changes to the compiler and delivering the with their macros... Posting this there is a great way to get those developers to join on discourse. The website is too static for fluent topics like this, in my opinion.

Reply:
I think the blog post is written for a general audience. Contributors don't need an explanation what Tasty is. So it fits better in the blog. We'll leave a separate thread on contributors.